File: mirror/tier2-php/showcase/pattern_exhaustive.php
<?php
// MIRROR V3: Tier 2 PHP Showcase
// Category: Pattern Matching
// Module: pattern_exhaustive.php

function exhaustive_match($value): string {
    if ($value === null) return 'nil';
    if (is_bool($value)) return 'boolean';
    if (is_int($value)) return 'integer';
    if (is_float($value)) return 'float';
    if (is_string($value)) return 'string';
    if (is_array($value)) return 'array';
    if (is_object($value)) return 'object';
    if (is_callable($value)) return 'function';
    return 'unknown';
}

function handle_all_types($value) {
    return match(exhaustive_match($value)) {
        'integer' => $value * 2,
        'string' => strtoupper($value),
        'array' => count($value),
        'object' => get_object_vars($value),
        'function' => 'callable',
        default => null,
    };
}

$result = [
    'type_nil' => exhaustive_match(null),
    'type_bool' => exhaustive_match(true),
    'type_int' => exhaustive_match(42),
    'type_float' => exhaustive_match(3.14),
    'type_str' => exhaustive_match('hello'),
    'type_list' => exhaustive_match([1, 2, 3]),
    'type_object' => exhaustive_match((object)['a' => 1]),
    'handle_int' => handle_all_types(10),
    'handle_str' => handle_all_types('test'),
];

echo json_encode($result);
?>

File: mirror/tier2-dart/showcase/pattern_exhaustive.dart
// MIRROR V3: Tier 2 Dart Showcase
// Category: Pattern Matching
// Module: pattern_exhaustive.dart

String exhaustiveMatch(dynamic value) {
  if (value == null) return 'nil';
  if (value is bool) return 'bool';
  if (value is int) return 'int';
  if (value is double) return 'double';
  if (value is String) return 'String';
  if (value is List) return 'List';
  if (value is Map) return 'Map';
  if (value is Function) return 'Function';
  return 'unknown';
}

dynamic handleAllTypes(dynamic value) {
  final typeStr = exhaustiveMatch(value);
  switch (typeStr) {
    case 'int': return (value as int) * 2;
    case 'String': return (value as String).toUpperCase();
    case 'List': return (value as List).length;
    case 'Map': return (value as Map).length;
    case 'Function': return 'callable';
    default: return null;
  }
}

void main() {
  final result = <String, dynamic>{
    'typeNil': exhaustiveMatch(null),
    'typeBool': exhaustiveMatch(true),
    'typeInt': exhaustiveMatch(42),
    'typeDouble': exhaustiveMatch(3.14),
    'typeStr': exhaustiveMatch('hello'),
    'typeList': exhaustiveMatch([1, 2, 3]),
    'typeMap': exhaustiveMatch({'a': 1}),
    'handleInt': handleAllTypes(10),
    'handleStr': handleAllTypes('test'),
  };
  print(result);
}
